### PR TITLE
An experiment on making higher threat rulesets more common.

### PR DESCRIPTION
**Pull request by ShiftyRail**
Without making rounds more gunked.

### Foreword

Yet another PR to throw numbers at dynamic. But we try to throw numbers in a *smart* way.
The key intent of this PR is to make traitor slightly less common, spicer gamemodes more common, without trying to slip back into "cultmalfrev" of early dynamic.

There's 3 changes.

### Changes

First, the *weight* of traitor-type rulesets has been lowered for midround and latejoin. (7 => 6)
**Reasoning :** This brings midround traitor and latejoin infiltrator more in line to roundstrart traitor. They shouldn't dwarf the other rulesets when dynamic picks a ruleset.

Secondly, the weight of a ruleset is reduced if a "sister" ruleset has been drafted is reduced a bit. If there's been traitor roundstart, perhaps it's better to send in another antag midround. Again, this is so midround traitor and latejoin traitor are no longer always bigger in weight than other rulesets.

Thirdly, and most importantly, if there's enough pop, and enough threat, instead of drafting more rulesets, there's a chance that "spicer" rulesets (the round-enders, which can't normally be drafted at the same time) will get a boost instead.

In short, if you've got a lot of threat and pop, there's less chance of "Vampire+Traitors" and more of "Nuclear Operatives" (alone).

:cl:
- tweak: Dynamic mode changes! (again!)
- tweak: Frequency of midround and latejoin traitor reduced, to give more room for other antag types.
- tweak: If the threat is high enough, there's a chance of boosting "round-ending" rulesets (like Revolution, Nuclear Operatives, Cult, AI malfunction...) instead of simply drafting more rulesets.

### CODE REVIEW

**Comment by Kurfursten** [comment]
Boosts **roundstart only malf, nuke emergency, revs, cult**
0-19 players: **effectively never boosts** (only exactly 100 threat)
20-24: 1 if 80+
25-30: 1 if 60+
31-35: 1 if 40+ (pretty common), but 2 if 70+ (pretty uncommon)
36-40: 1 if 20+ (very common), but 2 if 50+ (average)
41-44: Always +1, but 2 if 30+ (very common)
45-49: Always +1, but 2 if 10+ (almost always)
50+: Always +2

**Comment by Kurfursten** [comment]
I think one of the recurrent problems with Dynamic is not what it picks to begin with but because as time goes on if the round is *already peaceful* (it picked silent shufflers or generally nondestructive antags) it has no mechanism to get a louder round going.

In the first place, rulesets like revsquad or nuclear assault will be rejected because there aren't a lot of ghosts if no one is dead, and cult doesn't exist as a midround. At the same time, because these highlander rulesets have a higher threat cost, there's less chance of them being picked when some threat has already been spent on traitors.

**Comment by ShiftyRail** [comment]
I would argue that the failure to pick a "spicy" ruleset first is one of dynamic's weaknesses.

Also, this add the weight to *all* highlander rulesets, including midround ones.

**Comment by Kurfursten** [comment]
>Also, this add the weight to all highlander rulesets, including midround ones.

I don't think it adds weight to anything in this implementation now that I read it more closely. I think it creates a temporary instance of each subtype, called `rule`, and adds some weight to each of those but has no effect on any of the rulesets contained in `roundstart_rules`, `midround_rules` or `latejoin_rules`. Unless I'm misreading, that is.

**Comment by ShiftyRail** [comment]
You are correct
i'm dumb

**Comment by ShiftyRail** [comment]
Fixed

File: code/datums/gamemode/dynamic/dynamic.dm [hunk 1]
```diff
@@ -294,11 +285,28 @@ var/stacking_limit = 90
 
 	if	(extra_rulesets_amount && prob(50))
 		message_admins("Rather than extra rulesets, we'll try to draft spicier ones.")
-		for (var/datum/dynamic_ruleset/rule in subtypesof(/datum/dynamic_ruleset))
+		for (var/datum/dynamic_ruleset/rule in roundstart_rules)
```
Comment (Kurfursten) [inline code/datums/gamemode/dynamic/dynamic.dm:288]: still seems weird to have them increased here instead of inside get_weight
